In [ ]:
#| default_exp handlers.geotraces

# Geotraces

> This data pipeline, known as a "handler" in Marisco terminology, is designed to clean, standardize, and encode [BODC Geotraces dataset](https://www.geotraces.org/geotraces-intermediate-data-product-2021/) into MARIS `NetCDF` format. The handler processes `Geotraces` data, applying various transformations and lookups to align it with `MARIS` data standards.

Key functions of this handler:

- **Cleans** and **normalizes** raw Geotraces data
- **Applies standardized nomenclature** and units
- **Encodes the processed data** into `NetCDF` format compatible with MARIS requirements

This handler is a crucial component in the Marisco data processing workflow, ensuring Geotraces data is properly integrated into the MARIS database.

::: {.callout-tip}
## Getting Started

For new MARIS users, please refer to [MARIS Data Guide - A reference for data providers and data users](https://fr.anckalbi.net/marisco/reference/guide.html) for detailed information.

:::


The present notebook pretends to be an instance of [Literate Programming](https://www.wikiwand.com/en/articles/Literate_programming) in the sense that it is a narrative that includes code snippets that are interspersed with explanations. When a function or a class needs to be exported in a dedicated python module (in our case `marisco/handlers/geotraces.py`) the code snippet is added to the module using `#| exports` as provided by the wonderful [nbdev](https://nbdev.readthedocs.io/en/latest/) library.

In [ ]:
#| hide
from nbdev.showdoc import show_doc
%load_ext autoreload
%autoreload 2

In [ ]:
#| export
from fastcore.all import *
import pandas as pd
import numpy as np
import re

from marisco.callbacks import (
    Callback, 
    PerGroupCB,
    Transformer, 
    ParseTimeCB,
    SanitizeLonLatCB, 
    EncodeTimeCB,
    RemapCB
)

from marisco.metadata import (
    GlobAttrsFeeder, 
    BboxCB,
    DepthRangeCB, 
    TimeRangeCB,
    ZoteroCB,
    KeyValuePairCB
)

from marisco.configs import (
    AVOGADRO,
    get_lut,
    lut_path,
    cfg
)

from marisco.netcdf2csv import decode
from marisco.encoders import NetCDFEncoder

In [ ]:
#| hide
import warnings
pd.set_option('display.max_rows', 200)
warnings.filterwarnings('ignore')

## Configuration & file paths

- **fname_in**: path to the folder containing the HELCOM data in CSV format. The path can be defined as a relative path. 

- **fname_out**: path and filename for the NetCDF output.The path can be defined as a relative path. 

- **Zotero key**: used to retrieve attributes related to the dataset from [Zotero](https://www.zotero.org/). The MARIS datasets include a [library](https://maris.iaea.org/datasets) available on [Zotero](https://www.zotero.org/groups/2432820/maris/library). 


In [ ]:
# | exports
fname_in = '../../_data/geotraces/GEOTRACES_IDP2021_v2/seawater/ascii/GEOTRACES_IDP2021_Seawater_Discrete_Sample_Data_v2.csv'
fname_out = '../../_data/output/190-geotraces-2021.nc'
zotero_key = '97UIMEXN'

## Load data

In [ ]:
#| exports
load_data = lambda fname: pd.read_csv(fname_in)

In [ ]:
#| eval: false
df = load_data(fname_in)

df shape: (105417, 1188)


In [ ]:
print(f'df shape: {df.shape}')
df.columns

df shape: (105417, 1188)


Index(['Cruise', 'Station:METAVAR:INDEXED_TEXT', 'Type',
       'yyyy-mm-ddThh:mm:ss.sss', 'Longitude [degrees_east]',
       'Latitude [degrees_north]', 'Bot. Depth [m]',
       'Operator's Cruise Name:METAVAR:INDEXED_TEXT',
       'Ship Name:METAVAR:INDEXED_TEXT', 'Period:METAVAR:INDEXED_TEXT',
       ...
       'QV:SEADATANET.581', 'Co_CELL_CONC_BOTTLE [amol/cell]',
       'QV:SEADATANET.582', 'Ni_CELL_CONC_BOTTLE [amol/cell]',
       'QV:SEADATANET.583', 'Cu_CELL_CONC_BOTTLE [amol/cell]',
       'QV:SEADATANET.584', 'Zn_CELL_CONC_BOTTLE [amol/cell]',
       'QV:SEADATANET.585', 'QV:ODV:SAMPLE'],
      dtype='str', length=1188)

## Select columns of interest

The raw Geotraces CSV arrives in [**wide**](https://data.europa.eu/apps/data-visualisation-guide/wide-versus-long-data) format with 1,188 columns; mostly non-radionuclide parameters (nutrients, trace metals, quality flags) outside MARIS scope. The first step is to select only the radionuclide columns: `common_coi` lists the 6 metadata columns always kept as identifiers, and `nuclides_pattern` matches 80 measurement columns, reducing the table to 86. The regex patterns match on measurement column names, so companion quality-flag (`QV:`) columns are naturally excluded. The wide structure is then reshaped to long form in a later step.

In [ ]:
#| exports
# Metadata columns always kept as identifiers when reshaping wide → long
common_coi = ['yyyy-mm-ddThh:mm:ss.sss', 'Longitude [degrees_east]',
              'Latitude [degrees_north]', 'Bot. Depth [m]', 'DEPTH [m]', 'BODC Bottle Number:INTEGER']

# Regex patterns identifying radionuclide measurement columns
nuclides_pattern = ['^TRITI', '^Th_228', '^Th_23[024]', '^Pa_231', 
                    '^U_236_[DT]', '^Be_', '^Cs_137', '^Pb_210', '^Po_210',
                    '^Ra_22[3468]', '^Np_237', '^Pu_239_[D]', '^Pu_240', '^Pu_239_Pu_240',
                    '^I_129', '^Ac_227']  

In [ ]:
#| export
class SelectColsOfInterestCB(Callback):
    "Select columns of interest from the wide Geotraces dataframe."
    def __init__(self,
                 common_coi: list,       # Non-nuclide columns always kept as id_vars
                 nuclides_pattern: list  # Regex patterns matching nuclide column names
                 ): store_attr()
    def __call__(self, tfm):
        nuc_of_interest = [c for c in tfm.df.columns if 
                           any(re.match(pattern, c) for pattern in self.nuclides_pattern)]
        tfm.df = tfm.df[self.common_coi + nuc_of_interest]

For instance: 

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern)
])

df_test = tfm()

In [ ]:
print(f'First ten cols: {df_test.columns[:10]}')

# All metadata columns preserved
for col in common_coi: test_eq(col in df_test.columns, True)

# Quality flag columns stripped
test_eq(any('QV:' in c for c in df_test.columns), False)

First ten cols: Index(['yyyy-mm-ddThh:mm:ss.sss', 'Longitude [degrees_east]',
       'Latitude [degrees_north]', 'Bot. Depth [m]', 'DEPTH [m]',
       'BODC Bottle Number:INTEGER', 'TRITIUM_D_CONC_BOTTLE [TU]',
       'Cs_137_D_CONC_BOTTLE [uBq/kg]', 'I_129_D_CONC_BOTTLE [atoms/kg]',
       'Np_237_D_CONC_BOTTLE [uBq/kg]'],
      dtype='str')


**Columns matched by the nuclides patterns**  
From the 1,188 raw columns, the patterns above select these 80 radionuclide measurement columns:

In [ ]:
#| echo: false
# Actually load the data — just the column names
df = pd.read_csv(fname_in, nrows=0)  # just the header

nuclides_pattern = ['^TRITI', '^Th_228', '^Th_23[024]', '^Pa_231', 
                    '^U_236_[DT]', '^Be_', '^Cs_137', '^Pb_210', '^Po_210',
                    '^Ra_22[3468]', '^Np_237', '^Pu_239_[D]', '^Pu_240', '^Pu_239_Pu_240',
                    '^I_129', '^Ac_227']

for pat in nuclides_pattern:
    matched = [c for c in df.columns if re.match(pat, c)]
    print(f"  {pat:20s} → {len(matched):2d} cols  e.g. {matched[:3]}")
print(f"\nTotal nuclide columns selected: {sum(1 for c in df.columns if any(re.match(p,c) for p in nuclides_pattern))} of {len(df.columns)}")

  ^TRITI               →  1 cols  e.g. ['TRITIUM_D_CONC_BOTTLE [TU]']
  ^Th_228              →  4 cols  e.g. ['Th_228_D_CONC_PUMP [uBq/kg]', 'Th_228_D_CONC_UWAY [uBq/kg]', 'Th_228_SPT_CONC_PUMP [uBq/kg]']
  ^Th_23[024]          → 25 cols  e.g. ['Th_230_T_CONC_BOTTLE [uBq/kg]', 'Th_230_D_CONC_BOTTLE [uBq/kg]', 'Th_232_T_CONC_BOTTLE [pmol/kg]']
  ^Pa_231              →  8 cols  e.g. ['Pa_231_D_CONC_BOTTLE [uBq/kg]', 'Pa_231_D_CONC_FISH [uBq/kg]', 'Pa_231_D_CONC_UWAY [uBq/kg]']
  ^U_236_[DT]          →  3 cols  e.g. ['U_236_D_CONC_BOTTLE [atoms/kg]', 'U_236_T_CONC_BOTTLE [atoms/kg]', 'U_236_D_CONC_FISH [atoms/kg]']
  ^Be_                 →  2 cols  e.g. ['Be_7_T_CONC_PUMP [uBq/kg]', 'Be_7_D_CONC_PUMP [uBq/kg]']
  ^Cs_137              →  2 cols  e.g. ['Cs_137_D_CONC_BOTTLE [uBq/kg]', 'Cs_137_D_CONC_UWAY [uBq/kg]']
  ^Pb_210              →  7 cols  e.g. ['Pb_210_D_CONC_BOTTLE [mBq/kg]', 'Pb_210_D_CONC_FISH [mBq/kg]', 'Pb_210_D_CONC_UWAY [mBq/kg]']
  ^Po_210              →  7 cols  e.g. ['Po

## Reshape: wide to long

The raw Geotraces CSV is in wide format: each row holds up to 80 radionuclide
measurements crammed into separate columns, and metadata like unit, sampling
methodology, and filter status is embedded in the column names themselves (e.g.
`Th_230_D_CONC_BOTTLE [uBq/kg]`). This is unworkable for curation. Melting to
long format folds all measurements into a single `VALUE` column and a `NUCLIDE`
column that carries the full column-name string; which we can then parse to
extract unit, method, and filter status in the next step.

In [ ]:
#| export
class WideToLongCB(Callback):
    "Reshape wide nuclide columns to long format so unit, method, and filter status can be extracted from column names."
    def __init__(self,
                 common_coi: list,         # Non-nuclide columns kept as id_vars in melt
                 nuclides_pattern: list,   # Regex patterns identifying nuclide columns
                 var_name: str='NUCLIDE',  # Output column name for nuclide identifiers
                 value_name: str='VALUE',  # Output column name for measurement values
                 ): 
        store_attr()
        
    def __call__(self, tfm):
        nuc_of_interest = [c for c in tfm.df.columns if 
                           any(re.match(pattern, c) for pattern in self.nuclides_pattern)]
        tfm.df = pd.melt(tfm.df, id_vars=self.common_coi, value_vars=nuc_of_interest, 
                          var_name=self.var_name, value_name=self.value_name)
        tfm.df.dropna(subset=self.value_name, inplace=True)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern)
])
df_test = tfm()

In [ ]:
print(f'Long format: {df_test.shape[0]} rows × {df_test.shape[1]} cols')
# id columns all preserved
for col in common_coi: test_eq(col in df_test.columns, True)
# nuclide name and value columns created
test_eq('NUCLIDE' in df_test.columns, True)
test_eq('VALUE' in df_test.columns, True)
# no original wide nuclide columns remain
test_eq(any(re.match(p, c) for p in nuclides_pattern for c in df_test.columns), False)

Long format: 26745 rows × 8 cols


## Extract

Geotraces encodes **unit**, **filtering status**, and **sampling method** inside the column names themselves, for example `Th_230_D_CONC_BOTTLE [uBq/kg]` holds all three. These need to be parsed out into dedicated columns before they can drive unit conversion, MARIS nomenclature mapping, and quality checks.

### Unit

Units appear in square brackets at the end of every nuclide column name. The five distinct units found in this dataset are `uBq/kg`, `mBq/kg`, `TU`, `atoms/kg`, and `pmol/kg`; each needs to be mapped to the corresponding MARIS unit code in a later step.

In [ ]:
#| export
class ExtractUnitCB(Callback):
    "Extract measurement unit from nuclide column names (e.g. 'Cs_137_D_CONC_BOTTLE [uBq/kg]' → 'uBq/kg')."
    def __init__(self,
                 var_name: str='NUCLIDE'  # Column containing nuclide names with embedded units in brackets
                 ): 
        store_attr()
        self.unit_col_name = 'UNIT'

    def extract_unit(self, s):
        match = re.search(r'\[(.*?)\]', s)
        return match.group(1) if match else None
        
    def __call__(self, tfm):
        tfm.df[self.unit_col_name] = tfm.df[self.var_name].apply(self.extract_unit)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB()
])
df_test = tfm()

In [ ]:
print(f'Units found: {sorted(df_test.UNIT.unique())}')
test_eq('UNIT' in df_test.columns, True)
test_eq(set(df_test.UNIT.unique()), {'TU', 'uBq/kg', 'atoms/kg', 'mBq/kg', 'pmol/kg'})

Units found: ['TU', 'atoms/kg', 'mBq/kg', 'pmol/kg', 'uBq/kg']


### Filtering status

Phase codes embedded in nuclide column names encode both filtering status and sample type group. The second underscore component after the nuclide name indicates the phase: `D` (dissolved, FILT=1, SEAWATER), `T` (total, FILT=2, SEAWATER), and `TP` / `LPT` / `SPT` (suspended particulate matter fractions, all FILT=1, SUSPENDED_MATTER). These are parsed into dedicated `FILT` and `GROUP` columns that drive sample type classification and downstream quality checks.

In [ ]:
#| exports
# Phase code embedded in column names → FILT status and sample type group
phase = {
    'D': {'FILT': 1, 'group': 'SEAWATER'},
    'T': {'FILT': 2, 'group': 'SEAWATER'},
    'TP': {'FILT': 1, 'group': 'SUSPENDED_MATTER'}, 
    'LPT': {'FILT': 1, 'group': 'SUSPENDED_MATTER'},
    'SPT': {'FILT': 1, 'group': 'SUSPENDED_MATTER'}}

In [ ]:
#| export
class ExtractFilteringStatusCB(Callback):
    "Extract filtering status and sample-type group from nuclide column names using phase code (e.g. _D_, _T_, _TP_)."
    def __init__(self,
                 phase: dict,              # Phase code → {FILT, group} mapping (e.g. {'D': {'FILT': 1, 'group': 'SEAWATER'}})
                 var_name: str='NUCLIDE'   # Column containing nuclide names with embedded phase codes
                 ): 
        store_attr()
        self.filt_col_name = 'FILT'

    def extract_filt_status(self, s):
        matched_string = self.match(s)
        return self.phase[matched_string.group(1)][self.filt_col_name] if matched_string else None

    def match(self, s):
        return re.search(r'_(' + '|'.join(self.phase.keys()) + ')_', s)
        
    def extract_group(self, s):
        matched_string = self.match(s)
        return self.phase[matched_string.group(1)]['group'] if matched_string else None
        
    def __call__(self, tfm):
        tfm.df[self.filt_col_name] = tfm.df[self.var_name].apply(self.extract_filt_status)
        tfm.df['GROUP'] = tfm.df[self.var_name].apply(self.extract_group)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase)
])
df_test = tfm()

In [ ]:
print(f'Groups found: {sorted(df_test.GROUP.dropna().unique())}')
print(f'Filtering values: {sorted(df_test.FILT.dropna().unique())}')
test_eq('FILT' in df_test.columns, True)
test_eq('GROUP' in df_test.columns, True)
test_eq(set(df_test.GROUP.dropna().unique()), {'SEAWATER', 'SUSPENDED_MATTER'})
test_eq(set(df_test.FILT.dropna().unique()).issubset({1, 2}), True)

Groups found: ['SEAWATER', 'SUSPENDED_MATTER']
Filtering values: [np.int64(1), np.int64(2)]


### Sampling method

Sampling method codes appear as the last underscore component before the unit brackets in nuclide column names, for example `BOTTLE` in `Th_230_D_CONC_BOTTLE [uBq/kg]`. The four distinct methods found in this dataset are `BOTTLE` (rosette/CTD bottle, code 1), `FISH` (continuous towfish, code 18), `PUMP` (in situ pump, code 14), and `UWAY` (underway uncontaminated seawater supply, code 24). These are mapped to MARIS sampling method codes and recorded in the `SAMP_MET` column, enabling sample classification and cross dataset comparisons between different collection techniques.

In [ ]:
#| exports
# Sampling method code → MARIS method ID mapping (to be validated)
smp_method = {
    'BOTTLE': 1,
    'FISH': 18,
    'PUMP': 14,
    'UWAY': 24}

In [ ]:
#| export
class ExtractSamplingMethodCB(Callback):
    "Extract sampling method from nuclide names."
    def __init__(self, 
                 smp_method:dict = smp_method, # Sampling method lookup table
                 var_name='NUCLIDE',            # Column name containing nuclide names
                 smp_method_col_name = 'SAMP_MET' # Column name for sampling method in output df
                 ): 
        store_attr()

    def extract_smp_method(self, s):
        match = re.search(r'_(' + '|'.join(self.smp_method.keys()) + ') ', s)
        return self.smp_method[match.group(1)] if match else None
        
    def __call__(self, tfm):
        tfm.df[self.smp_method_col_name] = tfm.df[self.var_name].apply(self.extract_smp_method)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method)
])
df_test = tfm()

In [ ]:
print(f'Sampling methods found: {sorted(df_test.SAMP_MET.dropna().unique())}')
test_eq('SAMP_MET' in df_test.columns, True)
test_eq(set(df_test.SAMP_MET.dropna().unique()).issubset(set(smp_method.values())), True)

Sampling methods found: [np.int64(1), np.int64(14), np.int64(18), np.int64(24)]


## Remap to MARIS nuclide names 

Geotraces nuclide column names begin with provider-specific strings (e.g. `TRITIUM`, `Pu_239_Pu_240`, `Th_230`, `U_236`) that must be remapped to MARIS standard nomenclature before any lookup tables can be applied. Most names follow a regular pattern: strip the phase-code suffix (`_D`, `_T`, `_TP`, etc.), then lowercase and remove underscores — `Th_230` becomes `th230`, `U_236` becomes `u236`. Two exceptions need explicit overrides: `TRITIUM` maps to `h3` (the standard nuclide symbol for tritium), and `Pu_239_Pu_240` is a combined total activity that MARIS records as `pu239_240_tot`. The `RenameNuclideCB` applies the override dictionary first, then falls back to the general lowercasing rule for everything else.

In [ ]:
#| exports
# Provider-specific nuclide name overrides for MARIS standardisation
nuclides_name = {'TRITIUM': 'h3', 'Pu_239_Pu_240': 'pu239_240_tot'}

In [ ]:
#| export
class RenameNuclideCB(Callback):
    "Remap nuclides name to MARIS standard."
    def __init__(self,
                 nuclides_name: dict,     # Provider-specific name overrides e.g. {'TRITIUM': 'h3'}
                 var_name: str='NUCLIDE'  # Column containing nuclide names to standardize
                 ): 
        store_attr()
        self.patterns = ['_D', '_T', '_TP', '_LPT', '_SPT']

    def extract_nuclide_name(self, s):
        match = re.search(r'(.*?)(' + '|'.join(self.patterns) + ')', s)
        return match.group(1) if match else None

    def standardize_name(self, s):
        s = self.extract_nuclide_name(s)
        return self.nuclides_name[s] if s in self.nuclides_name else s.lower().replace('_', '')
        
    def __call__(self, tfm):
        tfm.df[self.var_name] = tfm.df[self.var_name].apply(self.standardize_name)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name)
])
df_test = tfm()

In [ ]:
nuclides = set(df_test.NUCLIDE.unique())
print(f'Nuclides after rename: {sorted(nuclides)}')
test_eq('h3' in nuclides, True)              # TRITIUM → h3 override
test_eq('pu239_240_tot' in nuclides, True)   # Pu_239_Pu_240 special case
test_eq(all(n == n.lower() for n in nuclides), True)  # all names lowercased

Nuclides after rename: ['ac227', 'be7', 'cs137', 'h3', 'i129', 'np237', 'pa231', 'pb210', 'po210', 'pu239', 'pu239_240_tot', 'pu240', 'ra223', 'ra224', 'ra226', 'ra228', 'th228', 'th230', 'th232', 'th234', 'u236']


In [ ]:
#| eval: false
df_test.NUCLIDE.unique()

<ArrowStringArray>
[           'h3',         'cs137',          'i129',         'np237',
         'pu239', 'pu239_240_tot',         'pu240',          'u236',
         'pa231',         'pb210',         'po210',         'ra224',
         'ra226',         'ra228',         'th230',         'th232',
         'th234',         'ac227',           'be7',         'ra223',
         'th228']
Length: 21, dtype: str

:::{.callout-important}
## FEEDBACK TO DATA PROVIDER

Several measurements are negative (see grouped counts below). Please review these values and provide detection-limit flags or handling guidance in future data releases.

:::

In [ ]:
#| eval: false
df_test[df_test.VALUE < 0].groupby('NUCLIDE').size()

NUCLIDE
h3       71
pa231     3
th228    22
th230     1
th232     6
dtype: int64

## Standardize unit

Geotraces encodes units inside nuclide column names, and five distinct
units appear across the dataset: TU, uBq/kg, mBq/kg, atoms/kg, and
pmol/kg. Some of these share a common MARIS unit ID despite different
magnitudes — uBq/kg and mBq/kg both map to unit ID 3 but differ by a
factor of 1000, which must be accounted for in the conversion factor.
Similarly, pmol/kg must be converted via Avogadro's number before it
matches the atoms/kg unit ID. The mapping below handles both the unit
remapping and the value rescaling:

In [ ]:
#| exports
# Geotraces unit → MARIS unit ID and conversion factor mapping
units_lut = {
    'TU': {'id': 7, 'factor': 1},
    'uBq/kg': {'id': 3, 'factor': 1e-6},
    'atoms/kg': {'id': 9, 'factor': 1},
    'mBq/kg': {'id': 3, 'factor': 1e-3},
    'pmol/kg': {'id': 9, 'factor': 1e-12 * AVOGADRO}
    }

In [ ]:
#| export
class StandardizeUnitCB(Callback):
    "Remap Geotraces unit strings to MARIS unit IDs, rescaling measurement values by the appropriate conversion factor where units share a common MARIS unit ID (e.g. uBq/kg and mBq/kg both map to ID 3 but differ 1000x)."
    def __init__(self, 
                 units_lut: dict,              # Unit string → {id, factor} conversion mapping
                 unit_col_name: str='UNIT',    # Column containing unit strings to remap
                 var_name: str='VALUE'         # Column containing measurement values to rescale
                 ): 
        store_attr()
        
    def __call__(self, tfm):
        # Convert/rescale values
        tfm.df[self.var_name] *= tfm.df[self.unit_col_name].map(
            {k: v['factor'] for k, v in self.units_lut.items()})
        
        # Match MARIS unit id
        tfm.df[self.unit_col_name] = tfm.df[self.unit_col_name].map(
            {k: v['id'] for k, v in self.units_lut.items()})

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut)
])
df_test = tfm()

In [ ]:
print(f'Unit IDs after standardization: {sorted(df_test.UNIT.unique())}')
test_eq(set(df_test.UNIT.unique()), {3, 7, 9})  # TU→7, uBq/kg+mBq/kg→3, atoms/kg+pmol/kg→9

Unit IDs after standardization: [np.int64(3), np.int64(7), np.int64(9)]


## Rename common columns

We rename the common columns to MARIS standard names before NetCDF encoding.

In [ ]:
#| exports
# Geotraces column name → MARIS standard name mapping
renaming_rules = {
    'yyyy-mm-ddThh:mm:ss.sss': 'TIME',
    'Longitude [degrees_east]': 'LON',
    'Latitude [degrees_north]': 'LAT',
    'DEPTH [m]': 'SMP_DEPTH',
    'Bot. Depth [m]': 'TOT_DEPTH',
    'BODC Bottle Number:INTEGER': 'SMP_ID_PROVIDER'
}

In [ ]:
#| export
class RenameColumnCB(Callback):
    "Renaming variables to MARIS standard names."
    def __init__(self,
                 lut: dict=renaming_rules  # Provider column name → MARIS standard name mapping
                 ): store_attr()
    def __call__(self, tfm):
        new_col_names = [self.lut[name] if name in self.lut else name for name in tfm.df.columns]
        tfm.df.columns = new_col_names

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules)
])
df_test = tfm()

In [ ]:
print(f'Columns after rename: {list(df_test.columns)}')
# MARIS standard names present
for new_name in renaming_rules.values(): test_eq(new_name in df_test.columns, True)
# Provider names removed
for old_name in renaming_rules.keys(): test_eq(old_name in df_test.columns, False)

Columns after rename: ['TIME', 'LON', 'LAT', 'TOT_DEPTH', 'SMP_DEPTH', 'SMP_ID_PROVIDER', 'NUCLIDE', 'VALUE', 'UNIT', 'FILT', 'GROUP', 'SAMP_MET']


## Unshift longitudes

In Geotraces, longitudes are coded between 0 and 360 in Geotraces. We rescale it between -180 and 180 instead.

In [ ]:
#| export
class UnshiftLongitudeCB(Callback):
    "Shift longitudes from Geotraces [0, 360] convention to MARIS [-180, 180] by subtracting 180."
    def __init__(self,
                 lon_col_name: str='LON'  # Column containing longitudes in [0, 360] to shift
                 ): 
        store_attr()
    def __call__(self, tfm):
        tfm.df[self.lon_col_name] = tfm.df[self.lon_col_name] - 180

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB()
])
df_test = tfm()

In [ ]:
print(f'LON range: [{df_test.LON.min():.4f}, {df_test.LON.max():.4f}]')
test_eq(df_test.LON.between(-180, 180).all(), True)

LON range: [-180.0000, 179.9986]


## Dispatch to groups

We encode each sample type (seawater, suspended matter, ...) into a dedicated dataframe as each sample type is further encoded as NetCDF group.

In [ ]:
#| export
class DispatchToGroupCB(Callback):
    "Split flat dataframe into per-group dict keyed by sample type (SEAWATER, SUSPENDED_MATTER, …)."
    def __init__(self,
                 group_name: str='GROUP'  # Column whose distinct values become the output dict keys
                 ): 
        store_attr()
        
    def __call__(self, tfm):
        tfm.dfs = dict(tuple(tfm.df.groupby(self.group_name)))
        for key in tfm.dfs:
            tfm.dfs[key] = tfm.dfs[key].drop(self.group_name, axis=1)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB()
])
dfs_test = tfm()

In [ ]:
print(f'Groups: {list(dfs_test.keys())}')
test_eq(set(dfs_test.keys()), {'SEAWATER', 'SUSPENDED_MATTER'})
# GROUP column consumed as dict key, not passed through
test_eq('GROUP' in dfs_test['SEAWATER'].columns, False)
test_eq('GROUP' in dfs_test['SUSPENDED_MATTER'].columns, False)

Groups: ['SEAWATER', 'SUSPENDED_MATTER']


## Add sample ID

After the wide-to-long reshape, each physical bottle (identified by `SMP_ID_PROVIDER`, renamed from `BODC Bottle Number:INTEGER`) appears once per nuclide. The counts below confirm this: fewer unique provider IDs than rows, so a row-level sequential `SMP_ID` is needed as the NetCDF dimension index.

In [ ]:
for grp, gdf in dfs_test.items():
    print(f'{grp}: {len(gdf)} rows, {gdf.SMP_ID_PROVIDER.nunique()} unique provider IDs')

SEAWATER: 19139 rows, 8779 unique provider IDs
SUSPENDED_MATTER: 7606 rows, 1849 unique provider IDs


The BODC Bottle Number is a stable provider identifier; it belongs in `SMP_ID_PROVIDER` for traceability. A sequential `SMP_ID` is generated per group to serve as the NetCDF dimension index.

In [ ]:
#| export
class AddSampleIDCB(PerGroupCB):
    "Assign a sequential SMP_ID per sample-type group; cast SMP_ID_PROVIDER (BODC Bottle Number) to string for NetCDF VLEN compatibility."
    def each_grp(self,
                 grp: str,          # Group key e.g. 'SEAWATER', 'SUSPENDED_MATTER'
                 df: pd.DataFrame,  # DataFrame for this group
                 tfm,               # Parent Transformer
                 ):
        df['SMP_ID'] = range(1, len(df) + 1)
        df['SMP_ID_PROVIDER'] = df['SMP_ID_PROVIDER'].astype(str)

In [ ]:
tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB(),
    AddSampleIDCB()
])
dfs_test = tfm()

In [ ]:
for grp, gdf in dfs_test.items():
    print(f'{grp}: SMP_ID range 1–{gdf.SMP_ID.max()}, SMP_ID_PROVIDER dtype={gdf.SMP_ID_PROVIDER.dtype}')
# SMP_ID is sequential from 1
test_eq(dfs_test['SEAWATER']['SMP_ID'].iloc[0], 1)
test_eq(dfs_test['SUSPENDED_MATTER']['SMP_ID'].iloc[0], 1)
# SMP_ID_PROVIDER cast to string for NetCDF VLEN compatibility
test_eq(dfs_test['SEAWATER']['SMP_ID_PROVIDER'].dtype, 'str')

SEAWATER: SMP_ID range 1–19139, SMP_ID_PROVIDER dtype=str
SUSPENDED_MATTER: SMP_ID range 1–7606, SMP_ID_PROVIDER dtype=str


## Parse time

We parse the time column to datetime format.

In [ ]:
#|eval: false
df = pd.read_csv(fname_in)

tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB(),
    ParseTimeCB()
])

dfs_test = tfm()
print('time data type: ', dfs_test['SEAWATER'].TIME.dtype)
print(dfs_test['SEAWATER'].head())

time data type:  datetime64[us]
                    TIME        LON      LAT  TOT_DEPTH  SMP_DEPTH  \
9223 2010-10-17 00:13:29  170.33792  38.3271     2827.0       17.8   
9231 2010-10-17 00:13:29  170.33792  38.3271     2827.0       34.7   
9237 2010-10-17 00:13:29  170.33792  38.3271     2827.0       67.5   
9244 2010-10-17 00:13:29  170.33792  38.3271     2827.0       91.9   
9256 2010-10-17 00:13:29  170.33792  38.3271     2827.0      136.6   

      SMP_ID_PROVIDER NUCLIDE  VALUE  UNIT  FILT  SAMP_MET  
9223           842525      h3  0.733     7     1         1  
9231           842528      h3  0.696     7     1         1  
9237           842531      h3  0.718     7     1         1  
9244           842534      h3  0.709     7     1         1  
9256           842540      h3  0.692     7     1         1  


## Encode time (seconds since ...)

Then encode it to seconds since `1970-01-01` as specified in MARIS NetCDF CDL and template. 


In [ ]:
#|eval: false
df = pd.read_csv(fname_in)

tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB(),    
    ParseTimeCB(),
    EncodeTimeCB()
])

dfs_test = tfm()['SEAWATER']
dfs_test.head()

,TIME,LON,LAT,TOT_DEPTH,SMP_DEPTH,SMP_ID_PROVIDER,NUCLIDE,VALUE,UNIT,FILT,SAMP_MET
9223,1287274409,170.33792,38.3271,2827.0,17.8,842525,h3,0.733,7,1,1
9231,1287274409,170.33792,38.3271,2827.0,34.7,842528,h3,0.696,7,1,1
9237,1287274409,170.33792,38.3271,2827.0,67.5,842531,h3,0.718,7,1,1
9244,1287274409,170.33792,38.3271,2827.0,91.9,842534,h3,0.709,7,1,1
9256,1287274409,170.33792,38.3271,2827.0,136.6,842540,h3,0.692,7,1,1


## Sanitize coordinates

In [ ]:
#|eval: false
df = pd.read_csv(fname_in)

tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB(),
    ParseTimeCB(),
    EncodeTimeCB(),
    SanitizeLonLatCB()
])
dfs_test = tfm()
dfs_test['SEAWATER'].head()

,TIME,LON,LAT,TOT_DEPTH,SMP_DEPTH,SMP_ID_PROVIDER,NUCLIDE,VALUE,UNIT,FILT,SAMP_MET
9223,1287274409,170.33792,38.3271,2827.0,17.8,842525,h3,0.733,7,1,1
9231,1287274409,170.33792,38.3271,2827.0,34.7,842528,h3,0.696,7,1,1
9237,1287274409,170.33792,38.3271,2827.0,67.5,842531,h3,0.718,7,1,1
9244,1287274409,170.33792,38.3271,2827.0,91.9,842534,h3,0.709,7,1,1
9256,1287274409,170.33792,38.3271,2827.0,136.6,842540,h3,0.692,7,1,1


## Remap nuclides name to id

All MARIS lookup tables are embeded in the NetCDF file as enumeration types. Data itself is encoded as integer for the sake of space efficiency. We need to remap it to the corresponding MARIS nuclide id.

In [ ]:
#| exports
# Lookup table: MARIS nc_name → nuclide_id
lut_nuclides = lambda: get_lut(lut_path(), 'dbo_nuclide.xlsx', 
                               key='nc_name', value='nuclide_id', reverse=False)

In [ ]:
lut_nuclides()

{'NOT APPLICABLE': -1,
 'NOT AVAILABLE': 0,
 'h3': 1,
 'be7': 2,
 'c14': 3,
 'k40': 4,
 'cr51': 5,
 'mn54': 6,
 'co57': 7,
 'co58': 8,
 'co60': 9,
 'zn65': 10,
 'sr89': 11,
 'sr90': 12,
 'zr95': 13,
 'nb95': 14,
 'tc99': 15,
 'ru103': 16,
 'ru106': 17,
 'rh106': 18,
 'ag106m': 19,
 'ag108': 20,
 'ag108m': 21,
 'ag110m': 22,
 'sb124': 23,
 'sb125': 24,
 'te129m': 25,
 'i129': 28,
 'i131': 29,
 'cs127': 30,
 'cs134': 31,
 'cs137': 33,
 'ba140': 34,
 'la140': 35,
 'ce141': 36,
 'ce144': 37,
 'pm147': 38,
 'eu154': 39,
 'eu155': 40,
 'pb210': 41,
 'pb212': 42,
 'pb214': 43,
 'bi207': 44,
 'bi211': 45,
 'bi214': 46,
 'po210': 47,
 'rn220': 48,
 'rn222': 49,
 'ra223': 50,
 'ra224': 51,
 'ra225': 52,
 'ra226': 53,
 'ra228': 54,
 'ac228': 55,
 'th227': 56,
 'th228': 57,
 'th232': 59,
 'th234': 60,
 'pa234': 61,
 'u234': 62,
 'u235': 63,
 'u238': 64,
 'np237': 65,
 'np239': 66,
 'pu238': 67,
 'pu239': 68,
 'pu240': 69,
 'pu241': 70,
 'am240': 71,
 'am241': 72,
 'cm242': 73,
 'cm243': 74,
 'cm24

In [ ]:
#|eval: false
df = pd.read_csv(fname_in)

tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB(),
    ParseTimeCB(),
    EncodeTimeCB(),
    SanitizeLonLatCB(),
    RemapCB(fn_lut=lut_nuclides, col_remap='NUCLIDE', col_src='NUCLIDE')
])

dfs_test = tfm()
dfs_test['SEAWATER'].NUCLIDE.unique()

Group BIOTA not found in the dataframes.
Group SEDIMENT not found in the dataframes.


array([  1,  33,  28,  65,  68,  77,  69, 108, 107,  41,  47,  51,  53,
        54, 106,  59,  60, 144,   2,  50,  57])

## NetCDF encoder

### Example change logs

In [ ]:
#|eval: false
df = pd.read_csv(fname_in)

tfm = Transformer(df, cbs=[
    SelectColsOfInterestCB(common_coi, nuclides_pattern),
    WideToLongCB(common_coi, nuclides_pattern),
    ExtractUnitCB(),
    ExtractFilteringStatusCB(phase),
    ExtractSamplingMethodCB(smp_method),
    RenameNuclideCB(nuclides_name),
    StandardizeUnitCB(units_lut),
    RenameColumnCB(renaming_rules),
    UnshiftLongitudeCB(),
    DispatchToGroupCB(),
    ParseTimeCB(),
    EncodeTimeCB(),
    SanitizeLonLatCB(),
    RemapCB(fn_lut=lut_nuclides, col_remap='NUCLIDE', col_src='NUCLIDE')
])

tfm();

Group BIOTA not found in the dataframes.
Group SEDIMENT not found in the dataframes.


In [ ]:
#|eval: false
tfm.logs

['Select columns of interest.',
 '\n    Get Geotraces nuclide names as values not column names \n    to extract contained information (unit, sampling method, ...).\n    ',
 '\n    Extract units from nuclide names.\n    ',
 'Extract filtering status from nuclide names.',
 'Extract sampling method from nuclide names.',
 'Remap nuclides name to MARIS standard.',
 'Remap unit to MARIS standard ones and apply conversion where needed.',
 'Renaming variables to MARIS standard names.',
 'Longitudes are coded between 0 and 360 in Geotraces. We rescale it between -180 and 180 instead.',
 'Convert to a dictionary of dataframe with sample type (seawater,...) as keys.',
 'Encode time as seconds since epoch.',
 'Drop rows with invalid longitude & latitude values. Convert `,` separator to `.` separator.',
 "Remap values from 'NUCLIDE' to 'NUCLIDE' for groups: dict_keys(['BIOTA', 'SEAWATER', 'SEDIMENT', 'SUSPENDED_MATTER'])."]

### Feed global attributes

In [ ]:
#| export
kw = ['oceanography', 'Earth Science > Oceans > Ocean Chemistry> Radionuclides',
      'Earth Science > Human Dimensions > Environmental Impacts > Nuclear Radiation Exposure',
      'Earth Science > Oceans > Ocean Chemistry > Ocean Tracers, Earth Science > Oceans > Marine Sediments',
      'Earth Science > Oceans > Ocean Chemistry, Earth Science > Oceans > Sea Ice > Isotopes',
      'Earth Science > Oceans > Water Quality > Ocean Contaminants',
      'Earth Science > Biological Classification > Animals/Vertebrates > Fish',
      'Earth Science > Biosphere > Ecosystems > Marine Ecosystems',
      'Earth Science > Biological Classification > Animals/Invertebrates > Mollusks',
      'Earth Science > Biological Classification > Animals/Invertebrates > Arthropods > Crustaceans',
      'Earth Science > Biological Classification > Plants > Macroalgae (Seaweeds)']

In [ ]:
#| exports
def get_attrs(
        tfm, 
        zotero_key, 
        kw=kw
        ):
    "Retrieve global attributes from Geotraces dataset."
    return GlobAttrsFeeder(tfm.dfs, cbs=[
        BboxCB(),
        DepthRangeCB(),
        TimeRangeCB(),
        ZoteroCB(zotero_key, cfg=cfg()),
        KeyValuePairCB('keywords', ', '.join(kw)),
        KeyValuePairCB('publisher_postprocess_logs', ', '.join(tfm.logs))
        ])()

In [ ]:
#|eval: false
zotero_metadata = get_attrs(tfm, zotero_key=zotero_key, kw=kw)
print('Keys: ', zotero_metadata.keys())
print('Title: ', zotero_metadata['title'])

Keys:  dict_keys(['geospatial_lat_min', 'geospatial_lat_max', 'geospatial_lon_min', 'geospatial_lon_max', 'geospatial_bounds', 'geospatial_vertical_max', 'geospatial_vertical_min', 'time_coverage_start', 'time_coverage_end', 'id', 'title', 'summary', 'creator_name', 'keywords', 'publisher_postprocess_logs'])
Title:  The GEOTRACES Intermediate Data Product 2017


### Encoding

In [ ]:
#| export
def encode(
        fname_in, 
        fname_out, 
        **kwargs
        ):
    df = pd.read_csv(fname_in)
    tfm = Transformer(df, cbs=[
        SelectColsOfInterestCB(common_coi, nuclides_pattern),
        WideToLongCB(common_coi, nuclides_pattern),
        ExtractUnitCB(),
        ExtractFilteringStatusCB(phase),
        ExtractSamplingMethodCB(smp_method),
        RenameNuclideCB(nuclides_name),
        StandardizeUnitCB(units_lut),
        RenameColumnCB(renaming_rules),
        UnshiftLongitudeCB(),
        DispatchToGroupCB(),
        AddSampleIDCB(),
        ParseTimeCB(),
        EncodeTimeCB(),
        SanitizeLonLatCB(),
        RemapCB(fn_lut=lut_nuclides, col_remap='NUCLIDE', col_src='NUCLIDE')
        ])
    
    tfm()
    encoder = NetCDFEncoder(tfm.dfs, 
                            # src_fname=nc_tpl_path,
                            dest_fname=fname_out,    
                            global_attrs=get_attrs(tfm, zotero_key=zotero_key, kw=kw),
                            verbose=kwargs.get('verbose', False)
                           )
    encoder.encode()

In [ ]:
#|eval: false
encode(fname_in, fname_out, verbose=False)

Group BIOTA not found in the dataframes.
Group SEDIMENT not found in the dataframes.


TODO:

- Add salinity, temperature, oxygen variables

In [ ]:
#|eval: false
decode(fname_in=fname_out, verbose=True)

Saved SEAWATER to ../../_data/output/190-geotraces-2021_SEAWATER.csv
Saved SUSPENDED_MATTER to ../../_data/output/190-geotraces-2021_SUSPENDED_MATTER.csv


When we use "#| exports" directive, it's ofen arguments (lut or other) we pass to the callback. According to DOC we have to add a comment before each to explain the purpose as for instance in 
```
#| exports
# Metadata columns always kept as identifiers when reshaping wide → long
common_coi = ['yyyy-mm-ddThh:mm:ss.sss', 'Longitude [degrees_east]',
              'Latitude [degrees_north]', 'Bot. Depth [m]', 'DEPTH [m]', 'BODC Bottle Number:INTEGER']

# Regex patterns identifying radionuclide measurement columns
nuclides_pattern = ['^TRITI', '^Th_228', '^Th_23[024]', '^Pa_231', 
                    '^U_236_[DT]', '^Be_', '^Cs_137', '^Pb_210', '^Po_210',
                    '^Ra_22[3468]', '^Np_237', '^Pu_239_[D]', '^Pu_240', '^Pu_239_Pu_240',
                    '^I_129', '^Ac_227']  